crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -297,102 +263,114 @@ impl TransactionBlock {
-                            query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
-                        }
-
-                        if let Some(o) = &filter.changed_object {
-                            let sub_query = tx_changed_objects::dsl::tx_changed_objects
-                                .select(tx_changed_objects::dsl::tx_sequence_number)
-                                .filter(tx_changed_objects::dsl::object_id.eq(o.into_vec()));
-                            query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
-                        }
-
-                        if let Some(txs) = &filter.transaction_ids {
-                            let digests: Vec<_> = txs.iter().map(|d| d.to_vec()).collect();
-                            query = query.filter(tx::dsl::transaction_digest.eq_any(digests));
-                        }
-
-                        query
-                    },
-                )
+                    scan_limit,
+                    &page,
+                )?
+                else {
+                    return Ok::<_, diesel::result::Error>((false, false, Vec::new(), None));
+                };
+
+                // If no filters are selected, or if the filter is composed of only checkpoint
+                // filters, we can directly query the main `transactions` table. Otherwise, we first
+                // fetch the set of `tx_sequence_number` from a join over relevant lookup tables,
+                // and then issue a query against the `transactions` table to fetch the remaining
+                // contents.
+                let (prev, next, transactions) = if !filter.has_filters() {
+                    let (prev, next, iter) = page.paginate_query::<StoredTransaction, _, _, _>(
+                        conn,
+                        checkpoint_viewed_at,
+                        move || {
+                            tx::transactions
+                                .filter(tx::tx_sequence_number.ge(tx_bounds.scan_lo() as i64))
+                                .filter(tx::tx_sequence_number.le(tx_bounds.scan_hi() as i64))
+                                .into_boxed()
+                        },
+                    )?;
+
+                    (prev, next, iter.collect())
+                } else {
+                    let subquery = subqueries(&filter, tx_bounds).unwrap();
+                    let (prev, next, results) =
+                        page.paginate_raw_query::<TxLookup>(conn, checkpoint_viewed_at, subquery)?;
+
+                    let tx_sequence_numbers = results
+                        .into_iter()
+                        .map(|x| x.tx_sequence_number)
+                        .collect::<Vec<i64>>();
+
+                    let transactions = conn.results(move || {
+                        tx::transactions
+                            .filter(tx::tx_sequence_number.eq_any(tx_sequence_numbers.clone()))
```

Comment:
nit: is the `clone` necessary?

Reply:
Yes I think so, because `conn.results` expects a query `Fn` which necessitates cloning `tx_sequence_numbers`, even though in our case we only use it once

Reply:
Ah yes! (We actually do call it multiple times -- once to do the EXPLAIN and once to execute it).